src/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/HostWorkspace/LanguageServerProjectLoader.cs

```diff
@@ -29,27 +32,70 @@ namespace Microsoft.CodeAnalysis.LanguageServer.HostWorkspace;
-    /// multiple projects at once; the key is a single List we just have a single thread processing any given project file. This is only to be used
-    /// in <see cref="LoadOrReloadProjectsAsync" /> and downstream calls; any other updating of this (like unloading projects) should be achieved by adding
-    /// things to the <see cref="ProjectsToLoadAndReload" />.
+    /// Guards access to <see cref="_loadedProjects"/>.
+    /// To keep the LSP queue responsive, <see cref="_gate"/> must not be held while performing design-time builds.
     /// </summary>
-    private readonly ConcurrentDictionary<string, List<LoadedProject>> _loadedProjects = [];
+    private readonly SemaphoreSlim _gate = new SemaphoreSlim(1);
+
+    /// <summary>
+    /// Guarded by <see cref="_gate"/>.
+    /// Maps the file path of a tracked project to the load state for the project.
+    /// Absence of an entry indicates the project is not tracked, e.g. it was never loaded, or it was unloaded.
+    /// <see cref="_gate"/> must be held when modifying the dictionary or objects contained in it.
+    /// </summary>
+    private readonly Dictionary<string, ProjectLoadState> _loadedProjects = [];
+
+    /// <summary>
+    /// State transitions:
+    /// <see cref="BeginLoading"/> -> <see cref="Loaded"/>
+    /// <see cref="BeginLoadingWithPrimordial"/> -> <see cref="Loaded"/>
+    /// Any state -> unloaded (which is denoted by removing the <see cref="_loadedProjects"/> entry for the project)
+    /// </summary>
+    private abstract record ProjectLoadState
+    {
+        private ProjectLoadState() { }
+
+        /// <summary>
+        /// The initial state used when we are asked to design-time build a project for the first time.
+        /// </summary>
+        internal sealed record BeginLoading : ProjectLoadState;
+
+        /// <summary>
+        /// Similar to <see cref="BeginLoading"/>, except including a <see cref="ProjectId"/> for a "primordial project".
+        /// </summary>
+        /// <param name="PrimordialProjectId">
+        /// ID of the project which LSP uses to fulfill requests until the first design-time build is complete.
+        /// The project with this ID is removed from the workspace when unloading or when transitioning to <see cref="Loaded"/> state.
+        /// </param>
+        internal sealed record BeginLoadingWithPrimordial(ProjectId PrimordialProjectId) : ProjectLoadState;
+
+        /// <summary>
+        /// The state after the first design-time build is finished.
+        /// Incrementally updated upon subsequent design-time builds.
+        /// The <see cref="LoadedProjectTargets"/> are disposed when unloading.
+        /// </summary>
+        /// <param name="LoadedProjectTargets">List of target frameworks which have been loaded for this project so far.</param>
+        internal sealed record Loaded(List<LoadedProject> LoadedProjectTargets) : ProjectLoadState;
```

Comment:
since you have records, could this be an IMmutableArray<LoadedProject> instead?  moving nodes through immutable states is nice.

Reply:
Yes, that should be simple for us to do.

Reply:
Done